### DISCUSSION TITLE
分布式训练cannot assign a device for operation PyFunc

### DISCUSSION THREAD

**Discussion by zhixiongning** [alibaba/GraphScope#1449] [Q&A] [answered]
在跑learning/examples/tf/graphsage/dist_train.py 时training过程报cannot assign device for operation PyFunc: {{node PyFunc} was explicitly assigned to /job:localhost/replica:0/task:0 but available devices are [/job:worker/replica:0:task:0/device:CPU:0]. Make sure the device specification refers to a valid device.
 我是在一台机器上启多个端口模拟分布式计算，未使用GPU

**Comment by sighingnow** [comment]
Please try to seek to tensorflow from similar issue reports (e.g., https://github.com/tensorflow/tensorflow/issues/31318 ) and see if it is helpful for your case.

**Comment by zhixiongning** [comment]
我参照相关issue尝试了一些办法还是没有解决。
另外，我参考这个https://github.com/tensorflow/tensorflow/issues/52959，修改了tensorflow的版本2.2.0(我之前用的最新版2.8.0)，但又遇到了grpc的错误。想问下当时这个分布式example是基于tensorflow的哪个版本运行的？谢谢

**Comment by sighingnow** [comment]
> 我参照相关issue尝试了一些办法还是没有解决。 另外，我参考这个[https://github.com/tensorflow/tensorflow/issues/52959，修改了tensorflow的版本2.2.0(我之前用的最新版2.8.0)，但又遇到了grpc的错误。想问下当时这个分布式example是基于tensorflow的哪个版本运行的？谢谢](https://github.com/tensorflow/tensorflow/issues/52959%EF%BC%8C%E4%BF%AE%E6%94%B9%E4%BA%86tensorflow%E7%9A%84%E7%89%88%E6%9C%AC2.2.0(%E6%88%91%E4%B9%8B%E5%89%8D%E7%94%A8%E7%9A%84%E6%9C%80%E6%96%B0%E7%89%882.8.0)%EF%BC%8C%E4%BD%86%E5%8F%88%E9%81%87%E5%88%B0%E4%BA%86grpc%E7%9A%84%E9%94%99%E8%AF%AF%E3%80%82%E6%83%B3%E9%97%AE%E4%B8%8B%E5%BD%93%E6%97%B6%E8%BF%99%E4%B8%AA%E5%88%86%E5%B8%83%E5%BC%8Fexample%E6%98%AF%E5%9F%BA%E4%BA%8Etensorflow%E7%9A%84%E5%93%AA%E4%B8%AA%E7%89%88%E6%9C%AC%E8%BF%90%E8%A1%8C%E7%9A%84%EF%BC%9F%E8%B0%A2%E8%B0%A2)

我们的CI中测试过的是tensorflow 2.5.2

**Comment by zhixiongning** [comment, marked as answer]
> > 我参照相关issue尝试了一些办法还是没有解决。 另外，我参考这个[https://github.com/tensorflow/tensorflow/issues/52959，修改了tensorflow的版本2.2.0(我之前用的最新版2.8.0)，但又遇到了grpc的错误。想问下当时这个分布式example是基于tensorflow的哪个版本运行的？谢谢](https://github.com/tensorflow/tensorflow/issues/52959%EF%BC%8C%E4%BF%AE%E6%94%B9%E4%BA%86tensorflow%E7%9A%84%E7%89%88%E6%9C%AC2.2.0(%E6%88%91%E4%B9%8B%E5%89%8D%E7%94%A8%E7%9A%84%E6%9C%80%E6%96%B0%E7%89%882.8.0)%EF%BC%8C%E4%BD%86%E5%8F%88%E9%81%87%E5%88%B0%E4%BA%86grpc%E7%9A%84%E9%94%99%E8%AF%AF%E3%80%82%E6%83%B3%E9%97%AE%E4%B8%8B%E5%BD%93%E6%97%B6%E8%BF%99%E4%B8%AA%E5%88%86%E5%B8%83%E5%BC%8Fexample%E6%98%AF%E5%9F%BA%E4%BA%8Etensorflow%E7%9A%84%E5%93%AA%E4%B8%AA%E7%89%88%E6%9C%AC%E8%BF%90%E8%A1%8C%E7%9A%84%EF%BC%9F%E8%B0%A2%E8%B0%A2)
> 
> 我们的CI中测试过的是tensorflow 2.5.2

谢谢，我使用2.2.0解决了这个问题
在import tensorflow之前我加了os.environ.pop('http_proxy')，os.environ.pop('https_proxy')

**Comment by sighingnow** [comment]
Close as resolved.
